### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [11]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head() #Display column headers for later reference

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44


## Player Count

* Display the total number of players


In [2]:
#Use a simple dataframe to print out a table similar to the example
plyr_tot = pd.DataFrame({"Total Players":[len(purchase_data["SN"].unique())]})
plyr_tot

Total Players
0            576

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
#Another simple dataframe for data and metadata
item_dat = purchase_data[["Purchase ID","Item Name","Price"]]
itm_res = pd.DataFrame({"Number of Unique Items":[len(item_dat["Item Name"].unique())],
                       "Average Price":[round(item_dat["Price"].mean(),2)],
                       "Number of Purchases":[item_dat["Purchase ID"].count()],
                       "Total Revenue":[item_dat["Price"].sum()]})
itm_res["Average Price"] = itm_res["Average Price"].map("${:.2f}".format)
itm_res["Total Revenue"] = itm_res["Total Revenue"].map("${:.2f}".format)
#Results
itm_res


Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
#New copied df that isolates values we want to analyze
#gndr - Gender
gndr_df = purchase_data[["SN","Gender"]].copy()

#Sort and remove duplicates
gndr_df.sort_values("SN",inplace = True)
gndr_df.drop_duplicates(keep = 'first', inplace = True)

#Results
res_gndr = pd.DataFrame(gndr_df["Gender"].value_counts())
res_gndr = res_gndr.rename(columns = {"Gender":"Total Count"})
res_gndr["Percentage of Players"] = round((res_gndr["Total Count"]/gndr_df["Gender"].count())*100,2)
res_gndr["Percentage of Players"] = res_gndr["Percentage of Players"].map("{:.2f}%".format)
res_gndr

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
#gp - Gender Purchase
gp_df = purchase_data[["SN","Gender","Price"]]
gp_group = gp_df.groupby("Gender")

#Results
res_gp = pd.DataFrame(gp_group["Price"].count())
res_gp["Average Purchase Price"] = round(gp_group["Price"].mean(),2)
res_gp["Total Purchase Value"] = round(gp_group["Price"].sum(),2)
res_gp["Avg Total Purchase per Person"] = round(res_gp["Total Purchase Value"]/res_gndr["Total Count"],2)

#Formatting
res_gp["Average Purchase Price"] = res_gp["Average Purchase Price"].map("${:.2f}".format)
res_gp["Total Purchase Value"] = res_gp["Total Purchase Value"].map("${:.2f}".format)
res_gp["Avg Total Purchase per Person"] = res_gp["Avg Total Purchase per Person"].map("${:.2f}".format)
res_gp

Price Average Purchase Price Total Purchase Value  \
Gender                                                                     
Female                   113                  $3.20              $361.94   
Male                     652                  $3.02             $1967.64   
Other / Non-Disclosed     15                  $3.35               $50.19   

                      Avg Total Purchase per Person  
Gender                                               
Female                                        $4.47  
Male                                          $4.07  
Other / Non-Disclosed                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
#Define bins for later
bin_age = [0,9,14,19,24,29,34,39,100]
bin_label = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

#Extracting values for analysis
age_df = purchase_data[["SN","Age"]].copy()
age_df.sort_values("SN",inplace = True)
age_df.drop_duplicates(keep = 'first', inplace = True)

#Results
res_age = pd.cut(age_df["Age"], bin_age, labels=bin_label,include_lowest = True).value_counts().sort_index()
res_age = pd.DataFrame(res_age).rename(columns = {"Age":"Total Count"})
res_age["Percentage of Players"] = round((res_age["Total Count"]/age_df["Age"].count())*100,2) 
res_age["Percentage of Players"] = res_age["Percentage of Players"].map("{:.2f}%".format)
res_age

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
#New extraction and cuts
agebuy_df = purchase_data[["SN","Age","Price"]].copy()
agebuy_df["Age Range"] = pd.cut(agebuy_df["Age"], bin_age, labels=bin_label,include_lowest = True)
age_group = agebuy_df.groupby(["Age Range"])

#Results
res_pa = pd.cut(agebuy_df["Age"], bin_age, labels=bin_label,include_lowest = True).value_counts().sort_index()
res_pa = pd.DataFrame(res_pa).rename(columns = {"Age":"Total Count"})
res_pa["Average Purchase"] = age_group["Price"].mean()
res_pa["Total Purchase"] = age_group["Price"].sum()
res_pa["Average Total Purchase"] = age_group["Price"].sum()/age_group["SN"].nunique()

#Formatting
res_pa["Average Purchase"] = res_pa["Average Purchase"].map("${:.2f}".format)
res_pa["Total Purchase"] = res_pa["Total Purchase"].map("${:.2f}".format)
res_pa["Average Total Purchase"] = res_pa["Average Total Purchase"].map("${:.2f}".format)
res_pa

Total Count Average Purchase Total Purchase Average Total Purchase
<10             23            $3.35         $77.13                  $4.54
10-14           28            $2.96         $82.78                  $3.76
15-19          136            $3.04        $412.89                  $3.86
20-24          365            $3.05       $1114.06                  $4.32
25-29          101            $2.90        $293.00                  $3.81
30-34           73            $2.93        $214.00                  $4.12
35-39           41            $3.60        $147.67                  $4.76
40+             13            $2.94         $38.24                  $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
#Extract SN data to distinguish who spends the most
sn_df = purchase_data[["SN","Price"]].copy()
sn_df.head()
sn_group = sn_df.groupby("SN")

#Results
res_sn = sn_group.count().rename(columns = {"Price":"Purchase Count"})
res_sn["Average Purchase Price"] = sn_group["Price"].mean()
res_sn["Total Purchase Value"] = sn_group["Price"].sum()
res_sn = res_sn.sort_values(["Total Purchase Value"], ascending=False)

#Formatting
res_sn["Average Purchase Price"] = res_sn["Average Purchase Price"].map("${:.2f}".format)
res_sn["Total Purchase Value"] = res_sn["Total Purchase Value"].map("${:.2f}".format)
res_sn.head() #Top 5 spenders

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
#Extract item information for analysis
item_df = purchase_data[["Item ID","Item Name","Price"]]
item_group = item_df.groupby(["Item ID","Item Name"])

#Results
res_item = item_group.count()
res_item["Count"] = item_group.count()
res_item["Price"] = item_group["Price"].mean()
res_item["Total Purchase Price"] = item_group["Price"].sum()
res_item = res_item.sort_values(["Count"], ascending=False)

#Formatting 
res_item["Price"] = res_item["Price"].map("${:.2f}".format)
res_item["Total Purchase Price"] = res_item["Total Purchase Price"].map("${:.2f}".format)
res_item.head()

Price  Count  \
Item ID Item Name                                                    
178     Oathbreaker, Last Hope of the Breaking Storm  $4.23     12   
145     Fiery Glass Crusader                          $4.58      9   
108     Extraction, Quickblade Of Trembling Hands     $3.53      9   
82      Nirvana                                       $4.90      9   
19      Pursuit, Cudgel of Necromancy                 $1.02      8   

                                                     Total Purchase Price  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
#Skip straight to results using the previous dataframes
res_itpr = item_group.count()
res_itpr["Price"] = item_group["Price"].mean()
res_itpr["Total Purchase Price"] = item_group["Price"].sum()
res_itpr = res_itpr.sort_values(["Total Purchase Price"], ascending=False)

#Formatting
res_itpr["Price"] = res_itpr["Price"].map("${:.2f}".format)
res_itpr["Total Purchase Price"] = res_itpr["Total Purchase Price"].map("${:.2f}".format)
res_itpr.head()

Price  \
Item ID Item Name                                             
178     Oathbreaker, Last Hope of the Breaking Storm  $4.23   
82      Nirvana                                       $4.90   
145     Fiery Glass Crusader                          $4.58   
92      Final Critic                                  $4.88   
103     Singed Scalpel                                $4.35   

                                                     Total Purchase Price  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

After analyzing the data, I came across the following results.

# Players

We have a total of 576 unique players playing our game! Across the 179 items that these players have bought, we have amassed a profit of \$2379.77 from their various purchases. I will dive deeper into these results below.

## Player Gender

Our current gender split has Male users leading at 84.03\%, Females at 14.06\%, and 1.91\% with undisclosed genders. Male users lead the pack in spending as well, spending a total of \$1967.64, though Female and Undisclosed users spend approximately 10\% more on average than Male users. 

## Player Age

Most of our players fall within the 20 to 24 y/o age range, accounting for 44.79\%, with the next highest groups being the 15 to 19 y/o and 25 to 29 y/o age ranges. Ideally, we should target this chunk together and market our product more to the 15 to 29 y/o age range as a whole if possible. As for spending, the 20-24 y/o age range spent a combined \\$1114.06 on over 300 in-game items, followed by the aforementioned 15-19 y/o and 25-29 y/o groups respectively. 

## Key Players

I also pulled results about some of our top spenders. Our current highest spender is **Lisosia93**. He has spent \\$18.96 on 5 different items. He's followed by **Idastidru52** (\\$15.45) and **Chamjask73** (\\$13.83).


# Items

On the topic of items, I decided to do a dive into the items that our users have been purchasing to see what's popular among our players. Currently, our **Oathbreaker, Last Hope of the Breaking Storm**	has net us \\$50.76 as our most popular and profitable item, followed by the **Fiery Glass Crusader** with a profit of \\$41.22 and **Extraction, Quickblade Of Trembling Hands** with a profit of \\$31.77. 